app/controllers/container_topology_controller.rb

```diff
@@ -0,0 +1,105 @@
+class ContainerTopologyController < ApplicationController
+  before_action :check_privileges
+  before_action :get_session_data
+  after_action :cleanup_action
+  after_action :set_session_data
+
+  def show
+    topology = build_topology
+    @topologyitems = topology[:items].to_json
+    @topologyrelations = topology[:relations].to_json
+    @topologykinds = topology[:kinds].to_json
+  end
+
+  def index
+    redirect_to :action => 'show'
+  end
+
+  def data
+    render :json => {:data => build_topology}
+  end
+
+  private
+
+  def get_session_data
+    @layout = "container_topology"
+  end
+
+  def set_session_data
+    session[:layout] = @layout
+  end
+
+  def build_topology
+    nodes, services = entities
+    topology = {}
+    topo_items = {}
+    links = []
+
+    nodes.each do |n|
+      topo_items[n.ems_ref] = {:metadata => {:id  => n.ems_ref, :name => n.name}, :kind => "Node"}
+      if n.container_groups.size > 0
+        n.container_groups.each do |cg|
+          topo_items[cg.ems_ref] = build_entity(cg, "ContainerGroup")
+          links << build_link(n, cg)
+          cg.containers.each do |c|
+            topo_items[c.ems_ref] = build_entity(c, "Container")
+            links << build_link(cg, c)
+          end
+        end
+
+      end
+      unless n.lives_on.nil?
+        kind = n.lives_on.type.end_with?('Vm') ? "VM" : "Host"
+        topo_items[n.lives_on.uid_ems] = {:metadata => {:id   => n.lives_on.uid_ems,
+                                                        :name => n.lives_on.name},
+                                          :kind     => kind}
+        links << {:source => n.ems_ref, :target => n.lives_on.uid_ems}
+        if kind == 'VM' # add link to Host
+          host = n.lives_on.host
+          topo_items[host.uid_ems] = {:metadata => {:id => host.uid_ems, :name => host.name}, :kind => 'Host'}
```

Comment:
> And here you can use the build_entity method.

I refactored the whole build_X methods thing to be able to use them everywhere through this code